share/src/main/java/org/apache/kafka/server/share/persister/PersisterStateManager.java

```diff
@@ -0,0 +1,863 @@
+                        return Collections.emptyList();
+                    }
+                    log.debug("Sending find coordinator RPC");
+                    return Collections.singletonList(new RequestAndCompletionHandler(
+                        time.milliseconds(),
+                        randomNode,
+                        handler.findShareCoordinatorBuilder(),
+                        handler
+                    ));
+                } else {
+                    // useful for tests and
+                    // other RPCs which might not be batchable
+                    if (!handler.isBatchable()) {
+                        requests.add(new RequestAndCompletionHandler(
+                            time.milliseconds(),
+                            handler.coordinatorNode,
+                            handler.requestBuilder(),
+                            handler
+                        ));
+                    }
+                }
+            }
+
+            // node1: {
+            //   group1: {
+            //      write: [w1, w2],
+            //      read: [r1, r2],
+            //      delete: [d1],
+            //      summary: [s1]
+            //   }
+            //   group2: {
+            //      write: [w3, w4]
+            //   }
+            // }
+            // For a sequence of writes, the flow would be:
+            // 1. 1st write request arrives
+            // 2. it is enqueued in the send thread
+            // 3. wakeup event causes the generate requests to find the coordinator
+            // 4. it will cause either RPC or cache lookup
+            // 5. once complete, the write handler is added to the nodeMap for batching and not the queue
+            // 6. wakeup event causes generate requests to iterate over the map and send the write request (W1) and
+            // remove node from the nodeMap and add it to inFlight
+            // 7. until W1 completes, more write requests (W2, W3, ...) could come in and get added to the nodeMap as per point 3, 4, 5.
+            // 8. if these belong to same node as W1. They will not be sent as the membership test with inFlight will pass.
+            // 9. when W1 completes, it will clear inFlight and raise wakeup event.
+            // 10. at this point W2, W3, etc. could be sent as a combined request thus achieving batching.
+            final Map<RPCType, Set<Node>> sending = new HashMap<>();
+            synchronized (nodeMapLock) {
+                nodeRPCMap.forEach((coordNode, rpcTypesPerNode) ->
+                    rpcTypesPerNode.forEach((rpcType, groupsPerRpcType) ->
+                        groupsPerRpcType.forEach((groupId, handlersPerGroup) -> {
+                            // this condition causes requests of same type and same destination node
+                            // to not be sent immediately but get batched
+                            if (!inFlight.containsKey(rpcType) || !inFlight.get(rpcType).contains(coordNode)) {
```

Comment:
1. The RPCs which need to be sent are NOT uniformly distributed by type and neither will they be in the future.
The bulk of the RPCs will by of write type. The reads will be far fewer since they will me made only when a SharePartition is initializing. Similar is the case for delete and initialize RPC in terms of volume.
Hence, increased batching on non-write RPCs will not have any noticeable effect.
2. Yes, because otherwise we might see starvation due to hot groups.